File: libs/frontend/domain/src/lib/model/conversation.ts
import {
  ConversationDescription,
  MessageDescription,
} from '../description/index.js';
import { Entity } from '../archtype/index.js';
import { Message } from './message.js';
import { HasManyPaged } from '../archtype/has-many.js';

export class Conversation implements Entity<string, ConversationDescription> {
  constructor(
    private identity: string,
    private description: ConversationDescription,
    private messages: ConversationMessages
  ) {}

  getIdentity(): string {
    return this.identity;
  }

  getDescription(): ConversationDescription {
    return this.description;
  }

  getMessages(): HasManyPaged<Message> {
    return this.messages;
  }

  saveMessage(description: MessageDescription): Promise<Message> {
    return this.messages.saveMessage(description);
  }
  sendMessage(
    message: string
  ): Promise<ReadableStream<Uint8Array<ArrayBuffer>>> {
    return this.messages.sendMessage(message);
  }
}

export interface ConversationMessages extends HasManyPaged<Message> {
  sendMessage(
    message: string
  ): Promise<ReadableStream<Uint8Array<ArrayBuffer>>>;
}

File: libs/backend/domain/src/main/java/com/businessdrivenai/domain/model/Conversation.java
package com.businessdrivenai.domain.model;

import com.businessdrivenai.archtype.Entity;
import com.businessdrivenai.archtype.HasMany;
import com.businessdrivenai.domain.description.ConversationDescription;
import com.businessdrivenai.domain.description.MessageDescription;
import jakarta.validation.constraints.NotNull;
import reactor.core.publisher.Flux;

public class Conversation implements Entity<String, ConversationDescription> {
  private String identity;
  private ConversationDescription description;
  private Messages messages;

  public Conversation(String identity, ConversationDescription description, Messages messages) {
    this.identity = identity;
    this.description = description;
    this.messages = messages;
  }

  public Conversation() {}

  @Override
  public String getIdentity() {
    return identity;
  }

  @Override
  public ConversationDescription getDescription() {
    return description;
  }

  public HasMany<String, Message> messages() {
    return messages;
  }

  public Message saveMessage(MessageDescription description) {
    return messages.saveMessage(description);
  }

  public Flux<String> sendMessage(
      MessageDescription description, ModelProvider modelProvider, String apiKey) {
    saveMessage(description);
    return modelProvider.sendMessage(description.content(), apiKey);
  }

  public interface Messages extends HasMany<String, Message> {
    Message saveMessage(MessageDescription description);
  }

  /** Interface for model providers, used to abstract different AI model services */
  public interface ModelProvider {
    /**
     * Send message to AI model and get response stream
     *
     * @param message User message content
     * @param apiKey API key for the AI model service
     * @return Response stream from AI model
     */
    Flux<String> sendMessage(String message, String apiKey);
  }

  public static class ConversationChange {
    @NotNull() private String title;

    public String getTitle() {
      return title;
    }

    public void setTitle(String title) {
      this.title = title;
    }
  }
}
